fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/JoinedRecordType.java

```diff
@@ -62,18 +62,25 @@ public static class Join {
         @Nonnull
         private final JoinConstituent left;
         @Nonnull
-        private final KeyExpression leftExpression;
+        private final KeyExpression leftSourceExpression;
+        @Nonnull
+        private final KeyExpression leftValueExpression;
         @Nonnull
         private final JoinConstituent right;
         @Nonnull
-        private final KeyExpression rightExpression;
+        private final KeyExpression rightSourceExpression;
+
+        @Nonnull
+        private final KeyExpression rightValueExpression;
 
-        protected Join(@Nonnull JoinConstituent left, @Nonnull KeyExpression leftExpression,
-                       @Nonnull JoinConstituent right, @Nonnull KeyExpression rightExpression) {
+        protected Join(@Nonnull JoinConstituent left, @Nonnull KeyExpression leftSourceExpression, @Nonnull KeyExpression leftValueExpression,
+                       @Nonnull JoinConstituent right, @Nonnull KeyExpression rightSourceExpression, @Nonnull KeyExpression rightValueExpression) {
```

Comment:
Could you go into a bit more detail into what the idea behind these new expressions is, and why they were necessary?

Reply:
If you want to join a string field to a numeric field, then you have to be able to supply a function key expression which can be applied when performing the join; that FKE is only useful when you are doing the join and shouldn't be used when actually writing the index, so you need two expressions--one for the index storage construction, and one to apply when performing the join

Reply:
Added documentation and changed variable names to reflect their relative roles